### PR TITLE
command-commands.lisp: execute-extended-command

### PR DESCRIPTION
**Pull request by jmercouris**
ability to execute an extended command with arguments :-)

How do we evaluate the arguments? Do we process an input of NIL as "NIL" or do we treat it as 'NIL?

### CODE REVIEW

File: source/command-commands.lisp [hunk 1]
```diff
@@ -41,6 +41,32 @@
       (setf (access-time command) (get-internal-real-time))
       (run command))))
 
+(define-command execute-extended-command ()
+  "Execute a command by name, also supply optional parameters."
+  (flet ((query-arguments (arguments command)
+               (flet ((make-binding (argument)
```
Comment (Ambrevar) [inline source/command-commands.lisp:47]: Indention issue?

File: source/command-commands.lisp [hunk 2]
```diff
@@ -41,6 +41,32 @@
...
+                        (let* ((sym (first argument))
+                               (default-value (format nil "~a" (second argument)))
```
Comment (Ambrevar) [inline source/command-commands.lisp:49]: In terms of style, I'm starting to wonder if we should replace all the `(format nil "~a" ...)` with `write-to-string`.  The latter is more readable and has the benefit of being more mapcar-able.

File: source/command-commands.lisp [hunk 3]
```diff
@@ -41,6 +41,32 @@
...
+                                   :input-buffer ,default-value))))))
+                 (let ((bindings (mapcar #'make-binding arguments))
+                       (bound-values (mapcar #'first arguments)))
+                   `(with-result* ,bindings
+                      (run ,command ,@bound-values))))))
+    (with-result (command (read-from-minibuffer
+                           (make-minibuffer
+                            :input-prompt "Execute command"
+                            :suggestion-function (command-suggestion-filter
+                                                  (mapcar (alex:compose #'class-name #'class-of)
```
Comment (Ambrevar) [inline source/command-commands.lisp:63]: Replace the composition by `#'mode-name`.

File: source/command-commands.lisp [hunk 4]
```diff
@@ -41,6 +41,32 @@
...
+                                  (make-minibuffer
+                                   :input-prompt ,string
+                                   :input-buffer ,default-value))))))
+                 (let ((bindings (mapcar #'make-binding arguments))
+                       (bound-values (mapcar #'first arguments)))
+                   `(with-result* ,bindings
+                      (run ,command ,@bound-values))))))
+    (with-result (command (read-from-minibuffer
+                           (make-minibuffer
+                            :input-prompt "Execute command"
```
Comment (Ambrevar) [inline source/command-commands.lisp:61]: Maybe "Execut extended command" or better "Execute command with arguments"?

**Comment by Ambrevar** [comment]
I would use a slightly different strategy: don't eval the whole call to (run...) but eval all argument individually.  This would be safer I think.

**Comment by Ambrevar** [comment]
About the `nil`, etc.  I think the answer is simple: use `read` :) 
This means that the user will have to surround strings with `"`.

File: source/command-commands.lisp [hunk 5]
```diff
@@ -41,6 +41,32 @@
...
+                            :suggestion-function (command-suggestion-filter
+                                                  (mapcar (alex:compose #'class-name #'class-of)
+                                                          (modes (current-buffer))))
+                            :hide-suggestion-count-p t)))
+      (setf (access-time command) (get-internal-real-time))
+      (let ((optional-arguments (nth-value 1 (alex:parse-ordinary-lambda-list 
+                                              (swank::arglist (sym command))))))
```
Comment (Ambrevar) [inline source/command-commands.lisp:68]: We also need the key arguments (nth-value 3 ...).
Many commands use them.

**Comment by Ambrevar** [comment]
Additionally, we could use the function type declaration (at least with SBCL) to infer the argument.
If the type is not found or too complex, then fall back to `read`.  We can modify the minibuffer prompt to let know the user when type is infered or not.

**Comment by Ambrevar** [comment]
You can get the function argument types with
`sb-introspect:function-type`.
See `describe-function` for example.

File: source/command-commands.lisp (hunk 5 above)
Comment (jmercouris) [reply source/command-commands.lisp:68]: Agreed, I just wanted to do a single step/first proof of concept :-)

File: source/command-commands.lisp (hunk 3 above)
Comment (jmercouris) [reply source/command-commands.lisp:63]: Shall I also change it in `EXECUTE-COMMAND`?

File: source/command-commands.lisp (hunk 4 above)
Comment (jmercouris) [reply source/command-commands.lisp:61]: `Execute extended command` should be OK. Eventually users will learn our terminology :-D

File: source/command-commands.lisp (hunk 2 above)
Comment (jmercouris) [reply source/command-commands.lisp:49]: Are they in fact equivalent? Will the aesthetic control character always produce write-to-string?

**Comment by jmercouris** [comment]
About the argument types, MAYBE. If you can find out how to do it cross implementation, I will be much more in favor!

File: source/command-commands.lisp (hunk 3 above)
Comment (jmercouris) [reply source/command-commands.lisp:63]: Changed in both locations.

File: source/command-commands.lisp (hunk 2 above)
Comment (Ambrevar) [reply source/command-commands.lisp]: If I get the specs right, the equivalent is princ-to-string.  

```
princ-to-string acts like write-to-string with :escape nil :readably nil. Thus no escape
characters are written. 

```

Example:

```lisp
(write-to-string "foo")
; "\"foo\""
(princ-to-string "foo")
; "foo"
```

So I suggest using `princ-to-string` when we want aesthetic, `write-to-string` when we want Lisp-readability.

**Comment by Ambrevar** [comment]
> About the argument types, MAYBE. If you can find out how to do it cross implementation, I will be much more in favor!

I could not find it for CCL.  In fact, declaiming the type in CCL does
not change the function description (e.g. `(describe #'foo)`).

That said, I'm in favour of adding this option for SBCL.  After all,
- it's the only compiler we officially support for now,
- it would not break CCL support,
- we already have a similar SBCL-specific feature in describe-function.

**Comment by jmercouris** [comment]
Can you please help me figure how to READ the results?

Let's say we have a command X Y Z.
In the code base, X Y and Z will be bound to:

```
(with-result* (x (read-from-minibuffer))
                    (y (read-from-minibuffer))
                    (z read-from-minibuffer))
  (some-body)
```
I then need to READ X Y and Z and then bind the values of (READ X) to X. For some reason or another, I cannot seem to figure this out...

**Comment by Ambrevar** [comment]
`(read-from-string x)` should parse X and return the correspond Lisp
form.

Ex:

```
(read-from-string "17")
; => 17

(read-from-string "(format t \"Hello ~a\" 17)")
; (FORMAT T "Hello ~a" 17)

**Comment by jmercouris** [comment]
My question was more along the lines of if there is a less convoluted way to achieve what I want rather than: a097e563b6e1f09a1449b0e2481dd71265f90535

File: source/command-commands.lisp (hunk 2 above)
Comment (jmercouris) [reply source/command-commands.lisp]: OK! :-)

**Comment by Ambrevar** [comment]
Why do you need to quote `read-from-string` ib the `mapcar`?

**Comment by jmercouris** [comment]
Because we need to take the result and read it from string, the minibuffer only returns string results.

**Comment by Ambrevar** [comment]
I tried to hack your function and much of the complexity seems to stem
from the "asynchronousness" of the minibuffer.

I suggest we first rework the minibuffer then give this a better shot.

**Comment by jmercouris** [comment]
Yes, hence why I am using a macro to generate the forms rather than just iterating through them :-D

**Comment by jmercouris** [comment]
Closing this pull request in favor of a new version using the synchronous minibuffer.

**Pull request closed without merging**